crypto/crmf/crmf_pbm.c

```diff
@@ -0,0 +1,222 @@
+        goto err;
+
+    /*
+     * owf identifies the hash algorithm and associated parameters used to
+     * compute the key used in the MAC process.  All implementations MUST
+     * support SHA-1.
+     */
+    if (!X509_ALGOR_set0(pbm->owf, OBJ_nid2obj(owfnid), V_ASN1_UNDEF, NULL)) {
+        error = CRMF_R_SETTING_OWF_ALGOR_FAILURE;
+        goto err;
+    }
+
+    /*
+     * iterationCount identifies the number of times the hash is applied
+     * during the key computation process.  The iterationCount MUST be a
+     * minimum of 100.      Many people suggest using values as high as 1000
+     * iterations as the minimum value.  The trade off here is between
+     * protection of the password from attacks and the time spent by the
+     * server processing all of the different iterations in deriving
+     * passwords.  Hashing is generally considered a cheap operation but
+     * this may not be true with all hash functions in the future.
+     */
+    if (itercnt < 100) {
+        error = CRMF_R_ITERATIONCOUNT_BELOW_100;
+        goto err;
+    }
+
+    if (!ASN1_INTEGER_set(pbm->iterationCount, itercnt))
+        goto err;
+
+    /*-
+     * mac identifies the algorithm and associated parameters of the MAC
+     * function to be used.  All implementations MUST support HMAC-SHA1 [HMAC].
+     * All implementations SHOULD support DES-MAC and Triple-DES-MAC [PKCS11].
+     */
+    if (!X509_ALGOR_set0(pbm->mac, OBJ_nid2obj(macnid), V_ASN1_UNDEF, NULL)) {
+        error = CRMF_R_SETTING_MAC_ALGOR_FAILURE;
+        goto err;
+    }
+
+    OPENSSL_free(salt);
+    return pbm;
+ err:
+    OPENSSL_free(salt);
+    OSSL_CRMF_PBMPARAMETER_free(pbm);
+    CRMFerr(CRMF_F_OSSL_CRMF_PBMP_NEW, error);
+    return NULL;
+}
+
+/*-
+ * calculates the PBM based on the settings of the given OSSL_CRMF_PBMPARAMETER
+ * |pbmp| identifies the algorithms, salt to use
+ * |msg| message to apply the PBM for
+ * |msglen| length of the message
+ * |sec| key to use
+ * |seclen| length of the key
+ * |mac| pointer to the computed mac, is allocated here,
+ *       will be freed if not pointing to NULL
+ * |maclen| pointer to the length of the mac, will be set
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CRMF_pbm_new(const OSSL_CRMF_PBMPARAMETER *pbmp,
+                      const unsigned char *msg, size_t msglen,
+                      const unsigned char *sec, size_t seclen,
+                      unsigned char **mac, unsigned int *maclen)
+{
+    int mac_nid, hmac_md_nid = NID_undef;
+    const EVP_MD *m = NULL;
+    EVP_MD_CTX *ctx = NULL;
+    unsigned char basekey[EVP_MAX_MD_SIZE];
+    unsigned int bklen = EVP_MAX_MD_SIZE;
+    int64_t iterations;
+    int error = CRMF_R_CRMFERROR;
+
+    if (mac == NULL || pbmp == NULL || pbmp->mac == NULL
+        || pbmp->mac->algorithm == NULL || msg == NULL || sec == NULL) {
+        error = CRMF_R_NULL_ARGUMENT;
+        goto err;
+    }
+    OPENSSL_free(*mac);
```

Comment:
I'm not sure this is wise. We don't own this memory and didn't allocate it. We have no business freeing it. Things might get confused if it wasn't allocated via OPENSSL_malloc().

Reply:
Good point; I've re-worked that.